src/app/views/builder/builder.component.ts

```diff
@@ -0,0 +1,201 @@
+  virtualPortalOutlet2: CdkPortalOutlet;
+  @ViewChild('virtualContainer3', { read: CdkPortalOutlet })
+  virtualPortalOutlet3: CdkPortalOutlet;
+
+  @ViewChild('ref3') ref3: ElementRef;
+  @ViewChild('ref4') ref4: ElementRef;
+
+  ngOnInit() {
+
+    this.subStyle = this.store.select(getStyle).subscribe((v) => {
+      this.currentControlItem = v;
+      this.droper = v.map((el) => el[0]);
+    });
+
+    this.subStyleGeneral = this.store.select(getGeneralStyle).subscribe((v) => {
+      this.stylesGeneral = v.stylesGeneral;
+      this.stylesGeneralInner = v.stylesGeneralInner;
+    });
+
+
+    this.form = new FormGroup({});
+  }
+
+  ngAfterViewInit() {
+    this.componentPortal = new ComponentPortal(StylePanelComponent);
+    this.virtualPortalOutlet.attach(this.componentPortal); // ERROR Error: NG0100 in console
+
+
+    this.domPortal = new DomPortal(this.ref3);
+    this.virtualPortalOutlet3.attach(this.domPortal);
+    this.domPortal = new DomPortal(this.ref4);
+    this.virtualPortalOutlet2.attach(this.domPortal);
+  }
+
+  drag = ['input', 'textarea', 'select', 'checkbox', 'button'];
+
+  droper = [];
+
+  drop(event: CdkDragDrop<any[]>): void {
+    if (event.previousContainer === event.container) {
+      moveItemInArray(
+        event.container.data,
+        event.previousIndex,
+        event.currentIndex
+      );
+    } else {
+      copyArrayItem(
+        event.previousContainer.data,
+        event.container.data,
+        event.previousIndex,
+        event.currentIndex
+      );
+
+      this.droper[event.currentIndex] =
+        this.droper[event.currentIndex] + '-' + this.droper.length;
+
+
+      let valueInput: any = { ...valueDefault };
+
+      let matchStyle;
+      switch (event.container.data[event.currentIndex].split('-')[0]) {
+        case 'input':
+          matchStyle = stylesSheet_Input;
+          break;
+        case 'checkbox':
+          matchStyle = stylesSheet_Checkbox;
+          break;
+        case 'select':
+          matchStyle = stylesSheet_Select;
+          valueInput.select = [
+            ['value1', 'Value1'],
+            ['value2', 'Value2'],
+          ];
+          break;
+        case 'textarea':
+          matchStyle = stylesSheet_Textarea;
+          break;
+        case 'button':
+          matchStyle = stylesSheet_Btn;
+          break;
+
+        default:
+          return console.log('Invalid type controll');
+      }
+
+      this.store.dispatch(
+        new AddStyleAction([
+          event.container.data[event.currentIndex],
+          matchStyle,
+          valueInput,
+        ])
+      );
+
+
+      this.currentControlItem.forEach((el) => {
+        let typeInput = el[0].split('-')[0];
+
+        let valueControl: any = '';
+
+        if (typeInput === 'button') {
+          return;
+        }
+        if (typeInput === 'checkbox') {
+          valueControl = false;
+        }
+
+        this.form.addControl(el[0], new FormControl(valueControl));
+      });
+    }
+  }
```

Comment:
refactor using SOLID and KISS principles

https://www.stijit.com/web-tips/dry-kiss-solid-yagni